File: src/tests/Interop/COM/NETClients/Primitives/ErrorTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

namespace NetClient
{
    using System;
    using System.Runtime.InteropServices;
    using Xunit;

    class ErrorTests
    {
        private readonly Server.Contract.Servers.ErrorMarshalTesting server;
        public ErrorTests()
        {
            this.server = (Server.Contract.Servers.ErrorMarshalTesting)new Server.Contract.Servers.ErrorMarshalTestingClass();
        }

        public void Run()
        {
            this.VerifyExpectedException();
            this.VerifyReturnHResult();
            this.VerifyHelpLink();
        }

        private void VerifyExpectedException()
        {
            Console.WriteLine($"Verify expected exception from HRESULT");

            Assert.Throws<NotImplementedException>(() => { this.server.Throw_HResult(unchecked((int)0x80004001)); });
            Assert.Throws<NullReferenceException>(() => { this.server.Throw_HResult(unchecked((int)0x80004003)); });
            Assert.Throws<UnauthorizedAccessException>(() => { this.server.Throw_HResult(unchecked((int)0x80070005)); });
            Assert.Throws<OutOfMemoryException>(() => { this.server.Throw_HResult(unchecked((int)0x8007000E)); });
            Assert.Throws<ArgumentException>(() => { this.server.Throw_HResult(unchecked((int)0x80070057)); });
            Assert.Throws<COMException>(() => { this.server.Throw_HResult(unchecked((int)0x8000ffff)); });
            Assert.Throws<COMException>(() => { this.server.Throw_HResult(unchecked((int)-1)); });
        }

        private void VerifyReturnHResult()
        {
            Console.WriteLine($"Verify preserved function signature");

            var hrs = new[]
            {
                unchecked((int)0x80004001),
                unchecked((int)0x80004003),
                unchecked((int)0x80070005),
                unchecked((int)0x80070057),
                unchecked((int)0x8000ffff),
                -1,
                1,
                2
            };

            foreach (var hr in hrs)
            {
                Assert.Equal(hr, this.server.Return_As_HResult(hr));
                Assert.Equal(hr, this.server.Return_As_HResult_Struct(hr).hr);
            }
        }

        private void VerifyHelpLink()
        {
            string helpLink = "X:\\NotA\\RealPath\\dummy.hlp";
            uint helpContext = 5678;
            var ex = Assert.Throws<COMException>(() => { this.server.Throw_HResult_HelpLink(unchecked((int)-1), helpLink, helpContext); });
            Assert.Equal($"{helpLink}#{helpContext}", ex.HelpLink);
        }
    }
}

File: src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#include "ClientTests.h"

namespace
{
    void VerifyExpectedException(_In_ IErrorMarshalTesting *et)
    {
       ::printf("Verify expected exception from HRESULT\n");

        HRESULT hrs[] =
        {
            E_NOTIMPL,
            E_POINTER,
            E_ACCESSDENIED,
            E_OUTOFMEMORY,
            E_INVALIDARG,
            E_UNEXPECTED,
            HRESULT{-1}
        };

        for (int i = 0; i < ARRAY_SIZE(hrs); ++i)
        {
            HRESULT hr = hrs[i];
            HRESULT hrMaybe = et->Throw_HResult(hr);
            THROW_FAIL_IF_FALSE(hr == hrMaybe);
        }
    }

    void VerifyReturnHResult(_In_ IErrorMarshalTesting *et)
    {
        ::printf("Verify preserved function signature\n");

        HRESULT hrs[] =
        {
            E_NOTIMPL,
            E_POINTER,
            E_ACCESSDENIED,
            E_INVALIDARG,
            E_UNEXPECTED,
            HRESULT{-1},
            S_FALSE,
            HRESULT{2}
        };

        for (int i = 0; i < ARRAY_SIZE(hrs); ++i)
        {
            HRESULT hr = hrs[i];
            HRESULT hrMaybe = et->Return_As_HResult(hr);
            THROW_FAIL_IF_FALSE(hr == hrMaybe);
        }
    }

    void VerifyReturnHResultStruct(_In_ IErrorMarshalTesting *et)
    {
        ::printf("Verify preserved function signature\n");

        HRESULT hrs[] =
        {
            E_NOTIMPL,
            E_POINTER,
            E_ACCESSDENIED,
            E_INVALIDARG,
            E_UNEXPECTED,
            HRESULT{-1},
            S_FALSE,
            HRESULT{2}
        };

        for (int i = 0; i < ARRAY_SIZE(hrs); ++i)
        {
            HRESULT hr = hrs[i];
            HRESULT hrMaybe = et->Return_As_HResult_Struct(hr);
            THROW_FAIL_IF_FALSE(hr == hrMaybe);
        }
    }

    void VerifyHelpContext(_In_ IErrorMarshalTesting *et)
    {
        ::printf("Verify expected helplink and context\n");

        HRESULT hrs[] =
        {
            E_NOTIMPL,
            E_POINTER,
            E_ACCESSDENIED,
            E_OUTOFMEMORY,
            E_INVALIDARG,
            E_UNEXPECTED,
            HRESULT{-1}
        };

        LPCWSTR helpLink = L"X:\\NotA\\RealPath\\dummy.hlp";

        for (int i = 0; i < ARRAY_SIZE(hrs); ++i)
        {
            HRESULT hr = hrs[i];
            DWORD helpContext = (DWORD)(i + 0x1234);
            HRESULT hrMaybe = et->Throw_HResult_HelpLink(hr, helpLink, helpContext);
            THROW_FAIL_IF_FALSE(hr == hrMaybe);

            ComSmartPtr<IErrorInfo> pErrInfo;
            THROW_IF_FAILED(GetErrorInfo(0, &pErrInfo));

            BSTR helpLinkMaybe;
            THROW_IF_FAILED(pErrInfo->GetHelpFile(&helpLinkMaybe));
            THROW_FAIL_IF_FALSE(TP_wcmp_s(helpLink, helpLinkMaybe) == 0);
            SysFreeString(helpLinkMaybe);

            DWORD helpContextMaybe;
            THROW_IF_FAILED(pErrInfo->GetHelpContext(&helpContextMaybe));
            THROW_FAIL_IF_FALSE(helpContext == helpContextMaybe);
        }
    }
}

void Run_ErrorTests()
{
    HRESULT hr;

    CoreShimComActivation csact{ W("NETServer"), W("ErrorMarshalTesting") };

    ComSmartPtr<IErrorMarshalTesting> errorMarshal;
    THROW_IF_FAILED(::CoCreateInstance(CLSID_ErrorMarshalTesting, nullptr, CLSCTX_INPROC, IID_IErrorMarshalTesting, (void**)&errorMarshal));

    VerifyExpectedException(errorMarshal);
    VerifyReturnHResult(errorMarshal);
    VerifyReturnHResultStruct(errorMarshal);
    VerifyHelpContext(errorMarshal);
}
